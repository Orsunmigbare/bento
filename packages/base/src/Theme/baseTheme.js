import { ThemeManager } from "@staccx/theme"

export const theme = {
  name: "Default theme",
  spacing: {
    huge: "96px",
    large: "48px",
    mediumPlus: "32px",
    medium: "24px",
    small: "12px",
    tiny: "6px",
    micro: "3px"
  },
  wrapper: {
    small: "420px",
    medium: "640px",
    large: "1024px",
    breakout: "640px"
  },
  color: {
    white: "#fff",
    bg: "#fff",
    subtleHover: "#F7F7FF",
    bgGray: "#f7f7ff",
    line: "#EBEBF2",
    gray: "#767676",
    wcag: "#777",
    disabled: "#BABABA",
    primary: "#00a190",
    secondary: "#F5A623",
    text: "#2b2b2b",
    negative: "#ea526f",
    positive: "#4A90E2",
    warning: "#c54059",
    black: "#000",
    green: "#00a190" // for toggle eks
  },
  fontFamily: {
    heading: "sans-serif",
    body: "sans-serif"
  },
  font: {
    h1: "36px",
    h2: "26px",
    h3: "22px",
    h4: "18px",
    h5: "16px",
    h6: "14px",
    base: "14px",
    tiny: "12px",
    input: "16px",
    huge: "44px"
  },
  targetSize: {
    small: "36px",
    normal: "44px"
  },
  borderRadius: "2px"
}

const manager = new ThemeManager(theme)
export default manager
