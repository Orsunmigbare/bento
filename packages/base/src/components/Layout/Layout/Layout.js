import React from "react"
import PropTypes from "prop-types"
import styled, { css } from "styled-components"
import { themify, spacing, wrapper } from "@staccx/theme"

const Layout = ({ grid, children, className }) => (
  <LayoutContainer className={className} grid={grid}>
    {children}
  </LayoutContainer>
)

export const LAYOUT = "LAYOUT"

const layoutStyle = layout => {
  switch (layout) {
    case "stonehenge":
      return css`
        grid-template-columns: 1fr minmax(320px, 1fr) minmax(320px, 1fr) 1fr;
        grid-template-areas:
          "header  header  header  header"
          ".  main  main  ."
          ".  body  aside  ."
          "footer  footer  footer  footer";
      `
    case "columnNarrow":
      return css`
        grid-template-columns: 1fr ${wrapper.small} 1fr;
        grid-template-areas:
          ".  header  ."
          ".  main  ."
          ".  body  ."
          ".  aside  ."
          ".  footer  .";
        grid-auto-flow: column;
      `
    default:
      break
  }
}

export const LayoutContainer = styled.div`
  display: grid;
  grid-row-gap: ${spacing.medium};
  grid-column-gap: ${spacing.medium};
  ${props => layoutStyle(props.grid)};
  ${themify(LAYOUT)};
`

Layout.defaultProps = {
  className: "",
  grid: "stonehenge"
}

Layout.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element
  ]).isRequired,
  className: PropTypes.string,
  grid: PropTypes.string
}

export default Layout
