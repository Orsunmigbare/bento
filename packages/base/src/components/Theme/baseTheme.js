export const theme = {
  name: "StaccXBase",
  spacing: {
    huge: "96px",
    large: "48px",
    mediumPlus: "32px",
    medium: "24px",
    small: "12px",
    tiny: "6px"
  },
  wrapper: {
    small: "420px",
    medium: "640px",
    large: "1024px"
  },
  color: {
    bg: "#fff",
    bgGray: "#f7f7ff",
    primary: "#00a190",
    gray: "#767676",
    text: "#2b2b2b",
    white: "#fff",
    red: "#ea526f",
    green: "#00a190",
    blue: "#988EFB",
    yellow: "#f4da87",
    line: "#EBEBF2",
    wcag: "#777",
    disabled: "#BABABA",
    warning: "#c54059",
    black: "#000",
    subtleHover: "#F7F7FF"
  },
  font: {
    size: {
      h1: "36px",
      h2: "26px",
      h3: "22px",
      h4: "18px",
      h5: "16px",
      h6: "14px",
      base: "14px",
      tiny: "12px",
      input: "16px",
      huge: "44px"
    },
    heading: "Libre Baskerville",
    body: "Libre Franklin"
  },
  targetSize: {
    small: "36px",
    normal: "44px"
  },
  borderRadius: "2px"
}

export default theme
