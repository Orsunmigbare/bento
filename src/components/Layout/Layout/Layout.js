import React from "react"
import PropTypes from "prop-types"
import styled, { css } from "styled-components"
import {
  spacing,
  grid,
  applyVariants,
  commonPropTypes,
  themePropTypes
} from "../../../theming"

/**
 * The default Layout-Component. Can be used with Both LayoutItem inside as well as other components.
 * Is very flexible. You can use some predefined grids with the grid-prop or you can use variants.
 * The default instance without props just add vertical spacing between items
 *
 */
const Layout = ({
  grid,
  children,
  className,
  variant,
  columnGap,
  rowGap,
  paddingTop,
  paddingBottom,
  ...restProps
}) => (
  <LayoutContainer
    className={className}
    grid={grid}
    variant={variant}
    columnGap={columnGap}
    rowGap={rowGap}
    paddingTop={paddingTop}
    paddingBottom={paddingBottom}
    {...restProps}
  >
    {children}
  </LayoutContainer>
)

Layout.themeProps = {
  container: {
    name: "LAYOUT",
    description: "",
    type: themePropTypes.style
  }
}

const isNotFlush = space => {
  switch (space) {
    case "flush":
      return false
    default:
      return true
  }
}

export const LayoutContainer = styled.div`
  display: grid;
  ${p =>
    isNotFlush(p.rowGap) &&
    css`
      grid-row-gap: ${spacing(p.rowGap)};
    `};
  ${p =>
    isNotFlush(p.columnGap) &&
    css`
      grid-column-gap: ${spacing(p.columnGap)};
    `};
  ${p =>
    isNotFlush(p.paddingBottom) &&
    css`
      padding-bottom: ${spacing(p.paddingBottom)};
    `};
  ${p =>
    isNotFlush(p.paddingTop) &&
    css`
      padding-top: ${spacing(p.paddingTop)};
    `};
  ${p => p.grid && grid};
  ${applyVariants(Layout.themeProps.container)};
`

Layout.defaultProps = {
  className: "",
  grid: null,
  columnGap: "medium",
  rowGap: "medium",
  paddingBottom: "flush",
  paddingTop: "flush"
}

Layout.propTypes = {
  /**
   * The component must have a child to structure the layout.
   */
  children: commonPropTypes.children.isRequired,
  /**
   * This prop assign a class to the component to style it.
   */
  className: PropTypes.string,
  /**
   * The prop accept columns and rows.
   */
  grid: PropTypes.string,
  /**
   * The spacing between columns.
   */
  columnGap: PropTypes.string,
  /**
   * The spacing between rows.
   */
  rowGap: PropTypes.string,
  /**
   * The padding-bottom of the component.
   */
  paddingBottom: PropTypes.string,
  /**
   * The padding-top of the component.
   */
  paddingTop: PropTypes.string
}

export default Layout
