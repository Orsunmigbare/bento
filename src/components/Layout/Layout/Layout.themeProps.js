import { themePropTypes } from "../../../theming"

const themeProps = {
  container: {
    name: "LAYOUT",
    description: "",
    type: themePropTypes.style
  }
}

export default themeProps
