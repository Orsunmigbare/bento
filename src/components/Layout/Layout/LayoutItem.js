import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import { applyVariants, commonPropTypes } from "../../../theming"
import themeProps from "./LayoutItem.themeProps"
import { componentCreateFactory } from "../../../theming/utils/createVariantsFunctionFactory"

const LayoutItem = ({ area, children, className, variant, ...rest }) => (
  <LayoutItemContainer
    className={className}
    area={area}
    variant={variant}
    {...rest}
  >
    {children}
  </LayoutItemContainer>
)

export const LayoutItemContainer = styled.div`
  grid-area: ${props => props.area};
  ${applyVariants(themeProps.container)};
`

LayoutItem.defaultProps = {
  className: "",
  area: "",
  variant: ""
}

LayoutItem.propTypes = {
  /**
   * The component require a child component.
   */
  children: commonPropTypes.children.isRequired,
  /**
   * This prop assign a class to the component to style it.
   */
  className: PropTypes.string,
  /**
   * The area of the component, e.g. header, body, aside or footer.
   */
  area: PropTypes.string,
  /**
   * The variant get assigned from the theme, so it basically is styling.
   */
  variant: PropTypes.string
}
LayoutItem.themeProps = themeProps
LayoutItem.createVariants = componentCreateFactory(LayoutItem)

export default LayoutItem
