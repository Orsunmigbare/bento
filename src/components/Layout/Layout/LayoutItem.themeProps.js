import { themePropTypes } from "../../../theming"

const themeProps = {
  container: {
    name: "LAYOUT_ITEM",
    description: "",
    type: themePropTypes.style
  }
}
export default themeProps
