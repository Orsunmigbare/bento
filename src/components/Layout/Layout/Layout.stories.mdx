import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import Layout from './Layout';
import LayoutItem from './LayoutItem';

<Meta title="Components|Layout/Layout and LayoutItem" component={Layout, LayoutItem}/>

# Layout and LayoutItem
This component can be used to create a page and structure it accordingly.

```javascript
import { Layout, LayoutItem } from "@staccx/bento"
```

## Standard Layout and LayoutItem
This is the standard Layout. The Layout component must have a child and it can work with any component. Below, the LayoutItem component is the child.

The Layout component can be styled with the props to create layouts with rows, columns and insert padding top and bottom. The grid prop accept following values: columns and rows.

The LayoutItem component has a prop to set the area, and the values are: header, body, aside and footer.

<Preview>
  <Story name="Standard Layout and LayoutItem">
    <div>
        <Layout 
            grid="rows" 
            style={{ backgroundColor: "#38618C", padding: 24, color: "white" }}>
            <LayoutItem 
                area="header"
                style={{ backgroundColor: "#FFFFFF", padding: 24, color: "black" }}>
                I'm the header inside a row!
            </LayoutItem>
        </Layout>
        <Layout 
            grid="columns"
            columnGap="large"
            style={{ backgroundColor: "#35A7FF", padding: 24, color: "white" }}
            >
            <LayoutItem 
                area="body"
                style={{ backgroundColor: "#FFFFFF", padding: 24, color: "black" }}>
                I'm the body inside a column!
            </LayoutItem>
            <LayoutItem 
                area="aside"
                style={{ backgroundColor: "#FFFFFF", padding: 24, color: "black" }}>
                I'm the aside inside a column!
            </LayoutItem>
        </Layout>
        <Layout 
            grid="rows" 
            style={{ backgroundColor: "#38618C", padding: 24, color: "white" }}>
            <LayoutItem 
                area="header"
                style={{ backgroundColor: "#FFFFFF", padding: 24, color: "black" }}>
                I'm the footer inside a row!
            </LayoutItem>
        </Layout>
    </div>
  </Story>
</Preview>

## Props for Layout
<Props of={Layout} />

## Props for LayoutItem
<Props of={LayoutItem} />