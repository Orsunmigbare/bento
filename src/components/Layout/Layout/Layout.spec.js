import React from "react"
import renderer from "react-test-renderer"
import { ThemeProvider } from "styled-components"
import baseTheme from "../../../theming/themes/baseTheme"
import Layout from "./Layout"
import LayoutItem from "./LayoutItem"

describe("Layout", () => {
  describe("Snapshots", () => {
    it("No props", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Layout>
              <LayoutItem>I'm the header inside a row!</LayoutItem>
            </Layout>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("ClassName", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Layout className="ClassName">
              <LayoutItem className="ClassName">
                I'm the header inside a row!
              </LayoutItem>
            </Layout>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("grid", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Layout grid="rows">
              <LayoutItem className="ClassName">
                I'm the header inside a row!
              </LayoutItem>
            </Layout>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("columnGap", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Layout columnGap="huge">
              <LayoutItem className="ClassName">
                I'm the header inside a row!
              </LayoutItem>
            </Layout>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("rowGap", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Layout rowGap="huge">
              <LayoutItem className="ClassName">
                I'm the header inside a row!
              </LayoutItem>
            </Layout>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("paddingBottom", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Layout PaddingBottom="huge">
              <LayoutItem className="ClassName">
                I'm the header inside a row!
              </LayoutItem>
            </Layout>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("paddingTop", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Layout PaddingTop="huge">
              <LayoutItem className="ClassName">
                I'm the header inside a row!
              </LayoutItem>
            </Layout>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("isNotFlush", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Layout rowGap="flush">
              <LayoutItem className="ClassName">
                I'm the header inside a row!
              </LayoutItem>
            </Layout>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("All", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Layout
              PaddingTop="huge"
              grid="rows"
              PaddingBottom="huge"
              rowGap="huge"
              columnGap="huge"
            >
              <LayoutItem className="ClassName">
                I'm the header inside a row!
              </LayoutItem>
            </Layout>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("No Children", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Layout />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("Item area", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Layout grid="rows">
              <LayoutItem className="ClassName" area="header">
                I'm the header inside a row!
              </LayoutItem>
            </Layout>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
