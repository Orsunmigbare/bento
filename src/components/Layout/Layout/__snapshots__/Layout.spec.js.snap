// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Snapshots All 1`] = `
<div
  className="sc-AykKC iSYdQW"
>
  <div
    className="sc-AykKD bHyrrD ClassName"
  >
    I'm the header inside a row!
  </div>
</div>
`;

exports[`Layout Snapshots ClassName 1`] = `
<div
  className="sc-AykKC bcKUMo ClassName"
>
  <div
    className="sc-AykKD bHyrrD ClassName"
  >
    I'm the header inside a row!
  </div>
</div>
`;

exports[`Layout Snapshots Item area 1`] = `
<div
  className="sc-AykKC dxRRlQ"
>
  <div
    className="sc-AykKD bUfcVE ClassName"
  >
    I'm the header inside a row!
  </div>
</div>
`;

exports[`Layout Snapshots No Children 1`] = `
<div
  className="sc-AykKC bcKUMo"
/>
`;

exports[`Layout Snapshots No props 1`] = `
<div
  className="sc-AykKC bcKUMo"
>
  <div
    className="sc-AykKD bHyrrD"
  >
    I'm the header inside a row!
  </div>
</div>
`;

exports[`Layout Snapshots columnGap 1`] = `
<div
  className="sc-AykKC hceKEx"
>
  <div
    className="sc-AykKD bHyrrD ClassName"
  >
    I'm the header inside a row!
  </div>
</div>
`;

exports[`Layout Snapshots grid 1`] = `
<div
  className="sc-AykKC dxRRlQ"
>
  <div
    className="sc-AykKD bHyrrD ClassName"
  >
    I'm the header inside a row!
  </div>
</div>
`;

exports[`Layout Snapshots isNotFlush 1`] = `
<div
  className="sc-AykKC fqZrUx"
>
  <div
    className="sc-AykKD bHyrrD ClassName"
  >
    I'm the header inside a row!
  </div>
</div>
`;

exports[`Layout Snapshots paddingBottom 1`] = `
<div
  className="sc-AykKC bcKUMo"
>
  <div
    className="sc-AykKD bHyrrD ClassName"
  >
    I'm the header inside a row!
  </div>
</div>
`;

exports[`Layout Snapshots paddingTop 1`] = `
<div
  className="sc-AykKC bcKUMo"
>
  <div
    className="sc-AykKD bHyrrD ClassName"
  >
    I'm the header inside a row!
  </div>
</div>
`;

exports[`Layout Snapshots rowGap 1`] = `
<div
  className="sc-AykKC iQdCeF"
>
  <div
    className="sc-AykKD bHyrrD ClassName"
  >
    I'm the header inside a row!
  </div>
</div>
`;
